internal/event/redis.go

```diff
@@ -39,93 +41,91 @@ func NewEventServiceRedis(logger logr.Logger, tp trace.TracerProvider, cfg *Conf
 		return nil, fmt.Errorf("instrument redis client tracing: %v", err)
 	}
 
-	return &EventServiceRedisImpl{
-		logger: logger,
-		client: client,
-		cfg:    cfg,
+	return &serviceRedisImpl[T]{
+		logger:     logger,
+		client:     client,
+		channel:    channel,
+		serializer: serializer,
 	}, nil
 }
 
-func (s *EventServiceRedisImpl) PublishEvent(ctx context.Context, event *goaingest.IngestEvent) {
+func (s *serviceRedisImpl[T]) PublishEvent(ctx context.Context, event T) {
 	ctx, cancel := context.WithTimeout(ctx, 1*time.Second)
 	defer cancel()
 
-	blob, err := json.Marshal(server.NewMonitorResponseBody(event))
+	blob, err := s.serializer.marshal(event)
 	if err != nil {
-		s.logger.Error(err, "Error encoding monitor event.")
+		s.logger.Error(err, "Error encoding event.")
+		return
 	}
 
-	if err := s.client.Publish(ctx, s.cfg.RedisChannel, blob).Err(); err != nil {
-		s.logger.Error(err, "Error publishing monitor event.")
+	if err := s.client.Publish(ctx, s.channel, blob).Err(); err != nil {
+		s.logger.Error(err, "Error publishing event.", "channel", s.channel)
 	}
 }
 
-func (s *EventServiceRedisImpl) Subscribe(ctx context.Context) (Subscription, error) {
-	sub := NewSubscriptionRedis(ctx, s.logger, s.client, s.cfg.RedisChannel)
-
+func (s *serviceRedisImpl[T]) Subscribe(ctx context.Context) (Subscription[T], error) {
+	sub := newSubscriptionRedis(ctx, s.logger, s.client, s.channel, s.serializer)
 	return sub, nil
 }
 
-// SubscriptionRedisImpl represents a stream of user-related events.
-type SubscriptionRedisImpl struct {
-	logger logr.Logger
-	pubsub *redis.PubSub
-	c      chan *goaingest.IngestEvent // channel of events
-	stopCh chan struct{}
+// subscriptionRedisImpl represents a stream of events.
+type subscriptionRedisImpl[T any] struct {
+	logger     logr.Logger
+	pubsub     *redis.PubSub
+	c          chan T
+	stopCh     chan struct{}
+	serializer eventSerializer[T]
 }
 
-var _ Subscription = (*SubscriptionRedisImpl)(nil)
-
-func NewSubscriptionRedis(
+func newSubscriptionRedis[T any](
 	ctx context.Context,
 	logger logr.Logger,
 	c redis.UniversalClient,
 	channel string,
-) Subscription {
+	serializer eventSerializer[T],
+) Subscription[T] {
 	pubsub := c.Subscribe(ctx, channel)
-	// Call Receive to force the connection to wait a response from
-	// Redis so the subscription is active immediately.
 	_, _ = pubsub.Receive(ctx)
-	sub := SubscriptionRedisImpl{
-		logger: logger,
-		pubsub: pubsub,
-		c:      make(chan *goaingest.IngestEvent, EventBufferSize),
-		stopCh: make(chan struct{}),
+	sub := subscriptionRedisImpl[T]{
+		logger:     logger,
+		pubsub:     pubsub,
+		c:          make(chan T, EventBufferSize),
+		stopCh:     make(chan struct{}),
+		serializer: serializer,
 	}
 	go sub.loop()
```

Comment:
Seems to me that we should pass `ctx` to `loop()`, and exit the loop if the context is cancelled. :thinking:

Reply:
Well spotted, I think this is a problem in my original implementation. The monitor handler is calling the Close method, but we could honour the ctx in the loop to ensure that the child goroutine is terminated if the monitor is stuck for some reason, which can be rare but it could happen. This can probably wait, but it'd be nice to do it at some point.